src/FakeItEasy/Creation/CastleDynamicProxy/CastleDynamicProxyGenerator.cs

```diff
@@ -179,9 +179,17 @@ private static ProxyGeneratorResult GetProxyResultForValueType(Type typeOfProxy)
                 interceptor);
         }
 
-        private static IEnumerable<Type> GetAllInterfacesToImplement(IEnumerable<Type> additionalInterfacesToImplement)
+        private static IEnumerable<Type> GetAllInterfacesToImplement(Type typeOfProxy, IEnumerable<Type> additionalInterfacesToImplement)
         {
-            return additionalInterfacesToImplement.Concat(typeof(ITaggable));
+            return additionalInterfacesToImplement
+                .Union(typeOfProxy.GetInterfaces().Where(CanInterfaceBeImplemented))
+                .Concat(typeof(ITaggable));
+        }
+
+        private static bool CanInterfaceBeImplemented(Type type)
+        {
+            var info = type.GetTypeInfo();
+            return (info.IsPublic || info.IsNestedPublic) && !info.IsImport;

```

Comment:
The use of `TypeInfo` rather than `Type`. I had initially used the exact code from Moq, but it didn't work on netstd.